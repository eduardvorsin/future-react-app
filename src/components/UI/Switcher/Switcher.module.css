.switcher {
	--width: 64px;
	--height: 34px;
	--gap: 2px;
	--radio-size: 28px;
	--border-width: 2px;

	margin: 0;
	width: var(--width);
	height: var(--height);
	border-width: var(--border-width);
	border-style: solid;
	border-color: var(--dark-color-100);
	display: inline-flex;
	border-radius: 18px;
	align-items: center;
	justify-content: space-between;
	padding: var(--gap);
	position: relative;
}

.switcher__label {
	/* styles for .switcher__label */
}

.switcher__radio {
	cursor: pointer;
	margin: 0;
	appearance: none;
	background-color: transparent;
	flex-grow: 1;
	position: relative;
	z-index: 5;
}

.switcher__status {
	position: absolute;
	top: 50%;
	left: 0;
	background-color: var(--dark-color-100);
	transform: translate(var(--gap), -50%);
	transition: left var(--transition-time) var(--transition-easing);
}

.switcher__radio,
.switcher__status {
	border-radius: 50%;
	width: var(--radio-size);
	height: var(--radio-size);
}

[data-theme='dark'] .switcher {
	border-color: var(--light-color-300);
}

[data-theme='dark'] .switcher__status {
	background-color: var(--light-color-300);
}

.switcher__radio:first-of-type:checked~.switcher__status {
	transform: translate(var(--gap), -50%);
	transition: transform var(--transition-time) var(--transition-easing);
}

.switcher__radio:last-of-type:checked~.switcher__status {
	--shift-size: calc(var(--width) - 100% - (var(--gap) * 2) - var(--border-width));

	transform: translate(var(--shift-size), -50%);
	transition: transform var(--transition-time) var(--transition-easing);
}