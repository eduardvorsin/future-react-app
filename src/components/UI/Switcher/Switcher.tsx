import React, { ChangeEventHandler, FC, useId } from 'react';
import classes from './Switcher.module.css';

type SwitcherProps = {
  className?: string,
  labelText: string
  checked: boolean,
  inputLabels: [string, string],
  name: string,
  value: [string, string],
  onChange: ChangeEventHandler<HTMLInputElement>,
  testId?: string,
}

const Switcher: FC<SwitcherProps> = ({
  className,
  labelText,
  checked,
  name,
  value,
  inputLabels,
  onChange,
  testId,
}) => {
  const id = useId();

  return (
    <fieldset
      className={`${classes.switcher} ${className}`}
      data-testid={testId}
    >
      <legend
        className={`${classes.switcher__label} sr-only`}
      >
        {labelText}
      </legend>
      <label
        className={`${classes['switcher__input-label']} sr-only`}
        htmlFor={`${id}-input-first`}
      >
        {inputLabels[0]}
      </label>
      <input
        id={`${id}-input-first`}
        className={classes.switcher__radio}
        name={name}
        role='switch'
        type='radio'
        value={value[0]}
        checked={checked}
        aria-checked={checked}
        onChange={onChange}
      />
      <label
        className={`${classes['switcher__input-label']} sr-only`}
        htmlFor={`${id}-input-second`}
      >
        {inputLabels[1]}
      </label>
      <input
        id={`${id}-input-second`}
        className={classes.switcher__radio}
        name={name}
        role='switch'
        type='radio'
        value={value[1]}
        checked={!checked}
        aria-checked={!checked}
        onChange={onChange}
      />
      <span className={classes.switcher__status}></span>
    </fieldset>
  );
};

export default Switcher;
