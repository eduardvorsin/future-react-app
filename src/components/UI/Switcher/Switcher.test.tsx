import React, { ChangeEvent } from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Switcher from './Switcher';

const mockInputLabels: [string, string] = ['on', 'off'];
const mockInputValues: [string, string] = ['on', 'off'];

describe('Switcher tests', () => {
  it('is rendered correctly', () => {
    const mockFn = jest.fn<void, [e: ChangeEvent]>();

    render(
      <Switcher
        checked
        name='test-switcher'
        labelText='test-switcher-label'
        inputLabels={mockInputLabels}
        value={mockInputValues}
        onChange={mockFn}
        testId='test-switcher'
      />,
    );

    expect(screen.getByTestId<HTMLFieldSetElement>('test-switcher')).toBeInTheDocument();
  });

  it('when the value changes, the mock function is triggered', async () => {
    const user = userEvent.setup();
    const mockFn = jest.fn<void, [e: ChangeEvent]>();

    render(
      <Switcher
        checked
        name='test-switcher'
        labelText='test-switcher-label'
        inputLabels={mockInputLabels}
        value={mockInputValues}
        onChange={mockFn}
        testId='test-switcher'
      />,
    );

    await user.click(screen.getAllByRole<HTMLInputElement>('switch')[1]);

    expect(mockFn).toHaveBeenCalledTimes(1);
  });

  it('is a basic snapshot', () => {
    const mockFn = jest.fn<void, [e: ChangeEvent]>();

    render(
      <Switcher
        checked
        name='test-switcher'
        labelText='test-switcher-label'
        inputLabels={mockInputLabels}
        value={mockInputValues}
        onChange={mockFn}
        testId='test-switcher'
      />,
    );

    expect(screen.getByTestId<HTMLFieldSetElement>('test-switcher')).toMatchSnapshot();
  });
});
