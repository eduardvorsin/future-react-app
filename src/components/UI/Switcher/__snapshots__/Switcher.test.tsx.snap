// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switcher tests is a basic snapshot 1`] = `
<fieldset
  class="switcher undefined"
  data-testid="test-switcher"
>
  <legend
    class="switcher__label sr-only"
  >
    test-switcher-label
  </legend>
  <label
    class="switcher__input-label sr-only"
    for=":r2:-input-first"
  >
    on
  </label>
  <input
    aria-checked="true"
    checked=""
    class="switcher__radio"
    id=":r2:-input-first"
    name="test-switcher"
    role="switch"
    type="radio"
    value="on"
  />
  <label
    class="switcher__input-label sr-only"
    for=":r2:-input-second"
  >
    off
  </label>
  <input
    aria-checked="false"
    class="switcher__radio"
    id=":r2:-input-second"
    name="test-switcher"
    role="switch"
    type="radio"
    value="off"
  />
  <span
    class="switcher__status"
  />
</fieldset>
`;
